### PR TITLE
Validate host constraints

### PR DESCRIPTION
**Pull request by NehaManjunath**

### CODE REVIEW

File: topo/node/node.go [hunk 1]
```diff
@@ -220,6 +225,66 @@ func (n *Impl) Create(ctx context.Context) error {
...
+}
+
+func (n *Impl) ValidateConstraints() []error {
+	hostConstraints := n.GetProto().GetHostConstraints()
+	if hostConstraints == nil {
+		return nil
+	}
+	var errorList []error
+	for _, hc := range hostConstraints {
+		switch v := hc.GetConstraint().(type) {
```
Comment (alexmasi) [inline topo/node/node.go:246]: add a log.V(1) here for each cons that we are validating

File: topo/node/node.go [hunk 2]
```diff
@@ -220,6 +225,66 @@ func (n *Impl) Create(ctx context.Context) error {
...
+	}
+	return combinedErrorString
+}
+
+func (n *Impl) ValidateConstraints() []error {
+	hostConstraints := n.GetProto().GetHostConstraints()
+	if hostConstraints == nil {
+		return nil
+	}
+	var errorList []error
```
Comment (alexmasi) [inline topo/node/node.go:244]: use the errlist library instead, should make this easier: https://github.com/openconfig/kne/blob/fd11a07f35a7afb429ecc511400c1f53e5ad3d6b/deploy/deploy.go#L174

File: topo/node/cisco/cisco.go [hunk 3]
```diff
@@ -405,6 +408,18 @@ func defaults(pb *tpb.Node) (*tpb.Node, error) {
 		if pb.Config.Image == "" {
 			pb.Config.Image = "xrd:latest"
 		}
+		if pb.HostConstraints == nil {
+			pb.HostConstraints = append(pb.HostConstraints,
+				&tpb.HostConstraint{Constraint: &tpb.HostConstraint_KernelConstraint{
+					KernelConstraint: &tpb.KernelParam{Name: "fs.inotify.max_user_instances",
+						ConstraintType: &tpb.KernelParam_BoundedInteger{BoundedInteger: &tpb.BoundedInteger{
+							MaxValue: 64000, MinValue: 0}}}}})
```
Comment (alexmasi) [inline topo/node/cisco/cisco.go:416]: I think this is a min cons, so leave max blank and it should default to infinity in the checking

File: topo/node/cisco/cisco.go [hunk 4]
```diff
@@ -405,6 +408,18 @@ func defaults(pb *tpb.Node) (*tpb.Node, error) {
...
+			pb.HostConstraints = append(pb.HostConstraints,
+				&tpb.HostConstraint{Constraint: &tpb.HostConstraint_KernelConstraint{
+					KernelConstraint: &tpb.KernelParam{Name: "kernel.pid_max",
+						ConstraintType: &tpb.KernelParam_BoundedInteger{BoundedInteger: &tpb.BoundedInteger{
+							MaxValue: 4194304, MinValue: 0}}}}})
```
Comment (alexmasi) [inline topo/node/cisco/cisco.go:421]: can leave min unspecified

File: topo/node/cisco/cisco.go [hunk 5]
```diff
@@ -405,6 +408,18 @@ func defaults(pb *tpb.Node) (*tpb.Node, error) {
 		if pb.Config.Image == "" {
 			pb.Config.Image = "xrd:latest"
 		}
+		if pb.HostConstraints == nil {
```
Comment (alexmasi) [inline topo/node/cisco/cisco.go:411]: also please confirm which constraint applies only to xrd or only to 8000e if that is the case

File: topo/node/cisco/cisco.go [hunk 6]
```diff
@@ -413,6 +423,13 @@ func defaults(pb *tpb.Node) (*tpb.Node, error) {
 		if pb.Config.Image == "" {
 			pb.Config.Image = "8000e:latest"
 		}
+		if pb.HostConstraints == nil {
```
Comment (alexmasi) [inline topo/node/cisco/cisco.go:426]: hostconstraints will never be nil at this point

File: topo/node/node.go [hunk 7]
```diff
@@ -220,6 +225,66 @@ func (n *Impl) Create(ctx context.Context) error {
...
+					errorList = append(errorList, fmt.Errorf("failed to validate kernel constraint error: %w", err))
+				}
+			}
+		}
+	}
+
+	return errorList
+}
+
+func validateBoundedInteger(protoConstraint *tpb.BoundedInteger, kernelConstraintInt int) error {
```
[the remark this answers is not in the document]
Comment (NehaManjunath) [reply topo/node/node.go]: Fixed.

File: topo/node/node.go (hunk 2 above)
Comment (NehaManjunath) [reply topo/node/node.go]: Done.

File: topo/node/cisco/cisco.go (hunk 4 above)
Comment (NehaManjunath) [reply topo/node/cisco/cisco.go]: Fixed.

File: topo/node/cisco/cisco.go (hunk 6 above)
Comment (NehaManjunath) [reply topo/node/cisco/cisco.go]: Fixed.

File: topo/node/cisco/cisco.go [hunk 8]
```diff
@@ -413,6 +438,13 @@ func defaults(pb *tpb.Node) (*tpb.Node, error) {
 		if pb.Config.Image == "" {
 			pb.Config.Image = "8000e:latest"
 		}
+		if len(pb.HostConstraints) == 0 {
```
Comment (alexmasi) [inline topo/node/cisco/cisco.go:441]: this condition will still never happen because of line 427

File: topo/node/node.go [hunk 9]
```diff
@@ -220,6 +228,79 @@ func (n *Impl) Create(ctx context.Context) error {
...
+			}
+		default:
+			return nil
+		}
+	}
+
+	return errorList.Err()
+}
+
+var (
```
Comment (alexmasi) [inline topo/node/node.go:262]: in the top var block I would define:

```
var (
    // Stubs for testing.
    kernelConstraintValue = kernelConstraintValueImpl
)

...

func kernelConstraintValueImpl(...) (int, error) {
    ...
}
```

This is a pattern we use for other stubs although nothing is wrong about your current approach.

File: topo/node/node.go [hunk 10]
```diff
@@ -220,6 +228,79 @@ func (n *Impl) Create(ctx context.Context) error {
...
+	kernelConstraintValue = func(basePath string, constraint string) (int, error) {
```
Comment (alexmasi) [inline topo/node/node.go:263]: why does this require a base path? Aren't all of the sysctl values found under the root dir?

File: topo/node/cisco/cisco.go (hunk 3 above)
Comment (NehaManjunath) [reply topo/node/cisco/cisco.go]: FIxed.

File: topo/node/cisco/cisco.go (hunk 8 above)
Comment (NehaManjunath) [reply topo/node/cisco/cisco.go:441]: line 427 is for model xrd. This line checks for hostconstraints in 8000e series based on a switch case. So this condition would happen, right?

File: topo/node/node.go (hunk 9 above)
Comment (NehaManjunath) [reply topo/node/node.go]: Thanks for the suggestion, modified accordingly.

File: topo/node/cisco/cisco.go (hunk 8 above)
Comment (NehaManjunath) [reply topo/node/cisco/cisco.go:441]: Resolved offline.

File: topo/node/node.go [hunk 11]
```diff
@@ -100,6 +105,24 @@ var (
...
+
+func kernelConstraintValueImpl(constraint string) (int, error) {
+	constraintData, err := os.ReadFile(convertSysctlNameToProcSysPath(constraint))
+	if err != nil {
+		return 0, err
+	}
+	kcValue, err := strconv.Atoi(strings.TrimSpace(string(constraintData)))
+	if err != nil {
+		return 0, fmt.Errorf("failed to convert kernel constraint data: %s error: %w",
+			string(constraintData), err)
```
Comment (alexmasi) [inline topo/node/node.go:121]: move these to the same line

File: topo/node/node.go [hunk 12]
```diff
@@ -220,6 +246,64 @@ func (n *Impl) Create(ctx context.Context) error {
...
+	return errorList.Err()
+}
+
+// validateBoundedInteger - Evaluates a constraint if is within a bound of max - min integer. It defaults any unspecified upper bound to infinity,
+// the lower bound should already default to zero if not specified, validates that lower <= upper in the constraint otherwise error
+func validateBoundedInteger(nodeConstraint *tpb.BoundedInteger, hostCons int) error {
+	if nodeConstraint.MinValue > nodeConstraint.MaxValue {
+		return fmt.Errorf("invalid bounds. Max value %d is less than min value %d", nodeConstraint.MaxValue, nodeConstraint.MinValue)
+	}
+	if !nodeConstraint.ProtoReflect().Has(nodeConstraint.ProtoReflect().Descriptor().Fields().ByName(protoreflect.Name("max_value"))) {
```
Comment (alexmasi) [inline topo/node/node.go:286]: i would just check if the field == 0 (the default value)

File: topo/node/cisco/cisco.go [hunk 13]
```diff
@@ -405,6 +423,13 @@ func defaults(pb *tpb.Node) (*tpb.Node, error) {
 		if pb.Config.Image == "" {
 			pb.Config.Image = "xrd:latest"
 		}
+		if len(pb.HostConstraints) == 0 {
```
Comment (alexmasi) [inline topo/node/cisco/cisco.go:426]: i think checking for nil is better here in case it is unspecified

**Review by alexmasi — COMMENTED** [review]
Okay, LGTM. Just 3 small nits and should be good to merge

File: topo/node/cisco/cisco.go (hunk 13 above)
Comment (NehaManjunath) [reply topo/node/cisco/cisco.go]: Addressed.

File: topo/node/node.go (hunk 10 above)
Comment (NehaManjunath) [reply topo/node/node.go]: My bad, had added it initially for testing. Removed it.

File: topo/node/cisco/cisco.go (hunk 5 above)
Comment (NehaManjunath) [reply topo/node/cisco/cisco.go:426]: Addressed. Constraints applicable per model:
xrd : fs.inotify.max_user_instances
8000e : kernel.pid_max

**Review by alexmasi — APPROVED** [review]

**Pull request merged**